cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,273 @@
+  private static final String SUBSCRIBER = "tether.server";
+  static final String TETHERING_TOPIC_PREFIX = "tethering_";
+  private final CConfiguration cConf;
+  private final TetherStore store;
+  private final MessagingService messagingService;
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  // Last processed message id for each topic. There's a separate topic for each peer.
+  private Map<String, String> lastMessageIds;
+
+  @Inject
+  TetherServerHandler(CConfiguration cConf, TetherStore store, MessagingService messagingService,
+                      TransactionRunner transactionRunner) {
+    this.cConf = cConf;
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+    initializeMessageIds();
+  }
+
+  /**
+   * Sends control commands to the client.
+   */
+  @GET
+  @Path("/tethering/controlchannels/{peer}")
+  public void connectControlChannel(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws IOException, NotImplementedException, PeerNotFoundException {
+    checkTetherServerEnabled();
+
+    store.updatePeerTimestamp(peer);
+    TetherStatus tetherStatus = store.getPeer(peer).getTetherStatus();
+    if (tetherStatus == TetherStatus.PENDING) {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+      return;
+    } else if (tetherStatus == TetherStatus.REJECTED) {
+      responder.sendStatus(HttpResponseStatus.FORBIDDEN);
+      return;
+    }
+
+    List<TetherControlMessage> commands = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), TETHERING_TOPIC_PREFIX + peer);
+    String lastMessageId = null;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, lastMessageIds.get(topic.getTopic()))) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetherControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                            TetherControlMessage.class);
+        commands.add(controlMessage);
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    }
+
+    if (lastMessageId != null) {
+      // Update the last message id for the topic if we read any messages
+      lastMessageIds.put(topic.getTopic(), lastMessageId);
+      String finalLastMessageId = lastMessageId;
+      TransactionRunners.run(transactionRunner, context -> {
```

Comment:
If you persist here and failed to send the commands later, then the messages will be lost?

Reply:
fixed. client now sends the message id in the request.